### ISSUE TITLE
Nanokernel FIFO does not handle timeouts correctly

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#153]
**_Reported by Peter Mitsis:_**

Unless I have missed something, I do not think that nanokernel FIFO's are handling timeouts correctly.  In the event that a fiber waiting on a FIFO times out, [stat] is not reverted.

For reference, the core of the nano_fifo structure is below.
struct nano_fifo {
    union {
        struct _nano_queue wait_q;
        struct _nano_queue data_a;
    };
    int stat;
};

If there is no data in the FIFO (stat <= 0), then a fiber that wants to get data from the FIFO will pend if timeout != TICKS_NONE.  Prior to pending, [stat] is decremented.  However, if the fiber times-out, [stat] is not reverted; this means that the FIFO object erroneously thinks that there is still a task waiting for data on the [wait_q].  This is particularly problematic when either a fiber or a task subsequently comes along and tries to add data to the nanokernel FIFO as it will erroneously try to remove a non-existent fiber from the [wait_q] and attempt to mark it as ready to run.

Steps to reproduce:
---------------------------
1. Define and initialize fifo
        struct nano_fifo  fifo;
        nano_fifo_init(&fifo);
2. Let a fiber wait on the fifo
        nano_fiber_fifo_get(&fifo, 2);
3. Let the fiber's request time out.
4. Add data to the FIFO.
        nano_task_fifo_put(&fifo, <data>);

If running on QEMU, QEMU will likely crash now.

(Imported from Jira ZEP-185)

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

The IP stack (exercised from the cc2520 driver) on Soletta over Zephyr also got these hangs recently (hang on nano_task_fifo_get(), in our case). It might be the same issue, and it's *not* resolved by this (https://gerrit.zephyrproject.org/r/#/c/1316/) patch.
I may open a separate issue for that, if it's preferable.

Here are the steps to reproduce:

- git clone https://github.com/solettaproject/soletta/
- cd into it and make thirdparty-update
- git clone https://github.com/solettaproject/soletta-small-oses-samples
- have a zephyr tree at HEAD (plus this patch)

So that the sol-fbp-generator tool is available as a host binary
while building for Zephyr, do

- build soletta on your host with
      make menuconfig
      goto "Basic"
      goto "Installation prefix"
      set it to somewhere you can install to (/tmp/soletta will do)
      exit Kconfig and make install

- export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/tmp/soletta/lib
- export PATH=$PATH:/tmp/soletta/bin

- git clean -xdf on Soletta, so that we can use it fresh for Zephyr again

- cd soletta-small-oses-samples/oic-client
- CFLAGS="-Os" make -C ../BUILD/ zephyr BOARD=quark_se_devboard KERNEL_TYPE=nano flash

(change the board, if applicable)

if you debug the running image, you'll see it hangs on a call to sol_socket_zephyr_sendmsg(), that leads to net_buf_get() -> nano_task_fifo_get() -> hang. This did not happen a while ago, we would just lose packets (another issue I'll have opened soon, but I need this figured out first).

**Comment by nashif** [comment]
**_by Peter Mitsis:_**

Gustavo,

I think a separate Jira issue should be filed for the issue you are encountering, as this particular Jira issue was originally filed to track one specific thing (now resolved by https://gerrit.zephyrproject.org/r/#/c/1316/) and that patch reportedly does not fix the issue you are seeing.

**Comment by nashif** [comment]
**_by Gustavo Lima Chaves:_**

Fair, Peter. Doing that.

**Comment by nashif** [comment]
**_by Peter Mitsis:_**

Merged into the code base on April 8-2016.

SHA1 ID: 5aaf7ab9309555c36b55cea719ad1ce20bedbbc4
Change-Id: Ifbcb5004558b06fc55cad2a955f5be20e716b392

**Comment by nashif** [comment]
**_by Sharron LIU:_**

We have test case to cover this scenario already. Verified passed.
```
# nano_fifo_init
# fiber_get_fifo_data_wait_timeout 100 ticks --> failed to get data
# let timeout arrival
# task_put_fifo_data
# fiber_get_fifo_data_wait_timeout a few ticks --> got data
```

**Comment by nashif** [comment]
**_by Sharron LIU:_**

We have test case to cover this scenario already. Verified passed.
```
# nano_fifo_init
# fiber_get_fifo_data_wait_timeout 100 ticks --> failed to get data
# let timeout arrival
# fiber_get_fifo_data_wait_timeout a few ticks --> got data
# task_put_fifo_data
```

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

Correcting the priority field
